datadog/resource_datadog_dashboard.go

```diff
@@ -0,0 +1,3856 @@
+	// Optional params
+	if datadogDefinition.Yaxis != nil {
+		terraformDefinition["yaxis"] = buildTerraformWidgetAxis(*datadogDefinition.Yaxis)
+	}
+	if datadogDefinition.Title != nil {
+		terraformDefinition["title"] = *datadogDefinition.Title
+	}
+	if datadogDefinition.TitleSize != nil {
+		terraformDefinition["title_size"] = *datadogDefinition.TitleSize
+	}
+	if datadogDefinition.TitleAlign != nil {
+		terraformDefinition["title_align"] = *datadogDefinition.TitleAlign
+	}
+	if datadogDefinition.Time != nil {
+		terraformDefinition["time"] = buildTerraformWidgetTime(*datadogDefinition.Time)
+	}
+	return terraformDefinition
+}
+
+func getHeatmapRequestSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		// A request should implement exactly one of the following type of query
+		"q":             getMetricQuerySchema(),
+		"apm_query":     getApmOrLogQuerySchema(),
+		"log_query":     getApmOrLogQuerySchema(),
+		"process_query": getProcessQuerySchema(),
+	}
+}
+func buildDatadogHeatmapRequests(terraformRequests *[]interface{}) *[]datadog.HeatmapRequest {
+	datadogRequests := make([]datadog.HeatmapRequest, len(*terraformRequests))
+	for i, _request := range *terraformRequests {
+		terraformRequest := _request.(map[string]interface{})
+		// Build HeatmapRequest
+		datadogHeatmapRequest := datadog.HeatmapRequest{}
+		if v, ok := terraformRequest["q"].(string); ok && len(v) != 0 {
+			datadogHeatmapRequest.SetMetricQuery(v)
+		} else if v, ok := terraformRequest["apm_query"].([]interface{}); ok && len(v) > 0 {
+			apmQuery := v[0].(map[string]interface{})
+			datadogHeatmapRequest.ApmQuery = buildDatadogApmOrLogQuery(apmQuery)
+		} else if v, ok := terraformRequest["log_query"].([]interface{}); ok && len(v) > 0 {
+			logQuery := v[0].(map[string]interface{})
+			datadogHeatmapRequest.LogQuery = buildDatadogApmOrLogQuery(logQuery)
+		} else if v, ok := terraformRequest["process_query"].([]interface{}); ok && len(v) > 0 {
+			processQuery := v[0].(map[string]interface{})
+			datadogHeatmapRequest.ProcessQuery = buildDatadogProcessQuery(processQuery)
+		}
+		datadogRequests[i] = datadogHeatmapRequest
+	}
+	return &datadogRequests
+}
+func buildTerraformHeatmapRequests(datadogHeatmapRequests *[]datadog.HeatmapRequest) *[]map[string]interface{} {
+	terraformRequests := make([]map[string]interface{}, len(*datadogHeatmapRequests))
+	for i, datadogRequest := range *datadogHeatmapRequests {
+		terraformRequest := map[string]interface{}{}
+		if datadogRequest.MetricQuery != nil {
+			terraformRequest["q"] = *datadogRequest.MetricQuery
+		} else if datadogRequest.ApmQuery != nil {
+			terraformQuery := buildTerraformApmOrLogQuery(*datadogRequest.ApmQuery)
+			terraformRequest["apm_query"] = []map[string]interface{}{terraformQuery}
+		} else if datadogRequest.LogQuery != nil {
+			terraformQuery := buildTerraformApmOrLogQuery(*datadogRequest.LogQuery)
+
```

Comment:
I'm not sure I fully follow the design of having one two functions per schema, one for buildDatadog and one for buildTerraform, mind elaborating on that?

Reply:
Sure! so for every resource, the buildTerraform function is used in read operations to build the Terraform resource data by syncing the local state with actual the upstream state (from `go-datadog-api`). The buildDatadog function is used in create/update operations to read the user's provided configuration in Terraform and maps them to the upstream API. 
I imagine this can get confusing to follow, but I think a good example to see the difference is to check `buildTerraformTemplateVariables` which accepts an array of `datadog.TemplateVariable` representing values from the upstream API and returns an array of string maps to be saved in the local state. `buildDatadogTemplateVariables` does the opposite by accepting an array of string maps representing user's configuration in Terrafrom and return an array of `datadog.TemplateVariable` objects to be pushed to the upstream API.

Reply:
I think this approach makes perfect sense. FTR, we're already using it e.g. in screenboard resource (there are `buildTF*` methods to properly fill in the `d` object and the rest `build*` methods which build the go-datadog-api objects that get sent to the API).

Reply:
Agree that makes sense. Thanks for the info!